File: src/application/sсhemas/product.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class SelectedOptionSchema(BaseModel):
    name: str
    value: str


class MoneySchema(BaseModel):
    amount: str
    currencyCode: str


class ProductVariantNodeSchema(BaseModel):
    id: str
    availableForSale: bool
    selectedOptions: list[SelectedOptionSchema]
    price: MoneySchema


class VariantEdgeSchema(BaseModel):
    node: ProductVariantNodeSchema


class ProductImageNodeSchema(BaseModel):
    url: str
    altText: str | None = None
    width: int | None = None
    height: int | None = None


class ImageEdgeSchema(BaseModel):
    node: ProductImageNodeSchema


class ProductOptionSchema(BaseModel):
    id: str
    name: str
    values: list[str]


class SeoSchema(BaseModel):
    title: str
    description: str


class PriceRangeSchema(BaseModel):
    minVariantPrice: MoneySchema
    maxVariantPrice: MoneySchema


class ProductBaseSchema(BaseModel):
    title: str = Field(..., max_length=100)
    description: str
    descriptionHtml: str | None = None
    handle: str | None = None


class ProductCreateSchema(ProductBaseSchema):
    price: float = Field(..., ge=0)
    options: list[ProductOptionSchema]
    tags: list[str] = []


class ProductResponseSchema(ProductBaseSchema):
    id: str  # Изменено на str для формата "product-{id}"
    availableForSale: bool
    priceRange: PriceRangeSchema
    options: list[ProductOptionSchema]
    variants: dict[str, list[VariantEdgeSchema]]  # GraphQL-style с edges
    featuredImage: ProductImageNodeSchema
    images: dict[str, list[ImageEdgeSchema]]  # GraphQL-style с edges
    seo: SeoSchema
    tags: list[str]
    updatedAt: datetime
    rating: float = Field(ge=0, le=5, default=0.0)
    # categoryId: int  # camelCase для единообразия

    model_config = ConfigDict(from_attributes=True)  # Для Pydantic v2


class ProductListResponseSchema(BaseModel):
    edges: list[dict[str, ProductResponseSchema]]  # GraphQL-style
    # Или для REST-style:
    # products: List[ProductResponseSchema]


# from pydantic import BaseModel, Field
# from typing import List, Optional
# from datetime import datetime

# class ProductOptionSchema(BaseModel):
#     id: str
#     name: str
#     values: List[str]

# class ProductVariantSchema(BaseModel):
#     id: str
#     available_for_sale: bool
#     selectedOptions: List[dict]
#     price: dict

# class ProductImageSchema(BaseModel):
#     url: str
#     altText: Optional[str] = None
#     width: Optional[int] = None
#     height: Optional[int] = None

# class ProductBaseSchema(BaseModel):
#     title: str = Field(..., max_length=100)
#     description: str
#     descriptionHtml: Optional[str] = None
#     handle: Optional[str] = None

# class ProductCreateSchema(ProductBaseSchema):
#     price: float = Field(..., ge=0)
#     options: List[ProductOptionSchema]
#     tags: List[str] = []

# class ProductResponseSchema(ProductBaseSchema):
#     id: int
#     available_for_sale: bool
#     priceRange: dict
#     variants: List[ProductVariantSchema]
#     featured_image: ProductImageSchema
#     images: List[ProductImageSchema]
#     seo: dict
#     tags: List[str]
#     updatedAt: datetime
#     rating: float = Field(ge=0, le=5, default=0.0)
#     category_id: int

#     class Config:
#         from_attributes = True

# class ProductListResponseSchema(BaseModel):
#     products: List[ProductResponseSchema]

# #-------------------

# # class VariantEdge(BaseModel):
# #     node: ProductVariantSchema

# # class ProductModel(BaseModel):
# #     variants: list[VariantEdge]
# #     images: list[ProductImageSchema]

File: frontend/lib/mappers/product.ts
/**
 * Type-safe mappers for converting between backend API types and frontend types
 */
import type { components } from '../api/types';
import { Product, ProductVariant, Image, Money } from 'lib/shopify/types';
import { getClientApiUrl } from '../config';

type BackendProduct = components['schemas']['ProductResponseSchema'];
type BackendMoney = components['schemas']['MoneySchema'];
type BackendImage = components['schemas']['ProductImageNodeSchema'];
type BackendVariant = components['schemas']['ProductVariantNodeSchema'];
type BackendVariantEdge = components['schemas']['VariantEdgeSchema'];
type BackendImageEdge = components['schemas']['ImageEdgeSchema'];
type BackendOption = components['schemas']['ProductOptionSchema'];
type BackendSelectedOption = components['schemas']['SelectedOptionSchema'];

// Simple backend product type that matches actual API response
interface SimpleBackendProduct {
  id: number;
  name: string;
  price: number;
  description: string;
  image_url?: string;
}

/**
 * Convert backend money format to frontend money format
 */
function mapMoney(backendMoney: BackendMoney): Money {
  return {
    amount: backendMoney.amount,
    currencyCode: backendMoney.currencyCode
  };
}

/**
 * Convert backend image format to frontend image format
 */
function mapImage(backendImage: BackendImage): Image {
  const apiBaseUrl = getClientApiUrl();

  // If the URL starts with '/', prepend the API base URL
  const imageUrl = backendImage.url.startsWith('/')
    ? `${apiBaseUrl}${backendImage.url}`
    : backendImage.url;

  return {
    url: imageUrl,
    altText: backendImage.altText || '',
    width: backendImage.width || undefined,
    height: backendImage.height || undefined
  };
}

/**
 * Convert backend variant format to frontend variant format
 */
function mapVariant(backendVariant: BackendVariant): ProductVariant {
  return {
    id: backendVariant.id,
    title: backendVariant.selectedOptions
      .map((opt: BackendSelectedOption) => opt.value)
      .join(' / ') || 'Default',
    availableForSale: backendVariant.availableForSale,
    selectedOptions: backendVariant.selectedOptions,
    price: mapMoney(backendVariant.price)
  };
}

/**
 * Convert simple backend product format (actual API response) to frontend product format
 */
export function mapSimpleBackendProductToFrontend(product: SimpleBackendProduct): Product {
  const apiBaseUrl = getClientApiUrl();

  return {
    id: product.id.toString(),
    handle: `product-${product.id}`,
    title: product.name || 'Product',
    description: product.description || '',
    descriptionHtml: product.description || '',
    updatedAt: new Date().toISOString(),
    tags: [],
    priceRange: {
      maxVariantPrice: {
        amount: product.price?.toString() || '0',
        currencyCode: 'USD'
      },
      minVariantPrice: {
        amount: product.price?.toString() || '0',
        currencyCode: 'USD'
      }
    },
    variants: [{
      id: `variant-${product.id}`,
      title: 'Default',
      availableForSale: true,
      selectedOptions: [],
      price: {
        amount: product.price?.toString() || '0',
        currencyCode: 'USD'
      }
    }],
    featuredImage: product.image_url ? {
      url: `${apiBaseUrl}${product.image_url}`,
      altText: product.name || 'Product image',
      width: 300,
      height: 300
    } : {
      url: '/placeholder.svg',
      altText: 'No image',
      width: 300,
      height: 300
    },
    images: product.image_url ? [{
      url: `${apiBaseUrl}${product.image_url}`,
      altText: product.name || 'Product image',
      width: 300,
      height: 300
    }] : [{
      url: '/placeholder.svg',
      altText: 'No image',
      width: 300,
      height: 300
    }],
    seo: {
      title: product.name || 'Product',
      description: product.description || ''
    },
    options: [],
    availableForSale: true
  };
}

/**
 * Map array of simple backend products to frontend products
 */
export function mapSimpleBackendProductsToFrontend(products: SimpleBackendProduct[]): Product[] {
  return products.map(mapSimpleBackendProductToFrontend);
}

/**
 * Convert backend product format to frontend product format
 */
export function mapBackendProductToFrontend(backendProduct: BackendProduct): Product {
  // Extract variants from the edge structure - handle case where variants might not exist
  const variants: ProductVariant[] = [];
  if (backendProduct.variants && typeof backendProduct.variants === 'object') {
    Object.values(backendProduct.variants).forEach((variantEdges: BackendVariantEdge[]) => {
      variantEdges.forEach((edge: BackendVariantEdge) => {
        variants.push(mapVariant(edge.node));
      });
    });
  }

  // Extract images from the edge structure - handle case where images might not exist
  const images: Image[] = [];
  if (backendProduct.images && typeof backendProduct.images === 'object') {
    Object.values(backendProduct.images).forEach((imageEdges: BackendImageEdge[]) => {
      imageEdges.forEach((edge: BackendImageEdge) => {
        images.push(mapImage(edge.node));
      });
    });
  }

  // Create default price range if not provided
  const defaultPrice = { amount: '0.00', currencyCode: 'USD' };
  const priceRange = backendProduct.priceRange ? {
    maxVariantPrice: mapMoney(backendProduct.priceRange.maxVariantPrice),
    minVariantPrice: mapMoney(backendProduct.priceRange.minVariantPrice)
  } : {
    maxVariantPrice: defaultPrice,
    minVariantPrice: defaultPrice
  };

  // Create default featured image if not provided
  const featuredImage = backendProduct.featuredImage ?
    mapImage(backendProduct.featuredImage) :
    { url: '', altText: '', width: undefined, height: undefined };

  return {
    id: backendProduct.id,
    handle: backendProduct.handle || backendProduct.id.toString(),
    title: backendProduct.title || 'Untitled Product',
    description: backendProduct.description || '',
    descriptionHtml: backendProduct.descriptionHtml || backendProduct.description || '',
    updatedAt: backendProduct.updatedAt || new Date().toISOString(),
    tags: backendProduct.tags || [],
    priceRange,
    variants,
    images,
    featuredImage,
    seo: backendProduct.seo ? {
      title: backendProduct.seo.title,
      description: backendProduct.seo.description
    } : {
      title: backendProduct.title || 'Product',
      description: backendProduct.description || ''
    },
    options: backendProduct.options ? backendProduct.options.map((option: BackendOption) => ({
      id: option.id,
      name: option.name,
      values: option.values
    })) : [],
    availableForSale: backendProduct.availableForSale ?? true
  };
}

/**
 * Map array of backend products to frontend products
 */
export function mapBackendProductsToFrontend(backendProducts: BackendProduct[]): Product[] {
  return backendProducts.map(mapBackendProductToFrontend);
}
